Terraform Plan and Apply GitHub Action Integrations (#401)
cloudposse/atmos: website/docs/integrations/github-actions/atmos-terraform-plan.md

---
title: Atmos Terraform Plan
sidebar_position: 40
sidebar_label: Terraform Plan
---

> The Cloud Posse GitHub Action "Atmos Terraform Plan" runs Terraform entirely from GitHub Action workflows.

Author: ```suggestion
The Cloud Posse GitHub Action for "Atmos Terraform Plan" runs Terraform entirely from GitHub Action workflows.
```

> Given any component and stack in an Atmos supported infrastructure environment, [`github-action-atmos-terraform-plan`](https://github.com/cloudposse/github-action-atmos-terraform-plan) will run `atmos terraform plan`, generate a Terraform planfile, store this planfile in a S3 Bucket with metadata in DynamodDB, and finally format the Terraform Plan result as part of a GitHub Workflow Summary.

Reviewer: ```suggestion
Given any component and stack in an Atmos supported infrastructure environment, [`github-action-atmos-terraform-plan`](https://github.com/cloudposse/github-action-atmos-terraform-plan) will run `atmos terraform plan`, generate a Terraform [planfile](https://developer.hashicorp.com/terraform/tutorials/automation/automate-terraform), store this planfile in an S3 Bucket with metadata in DynamodDB, and finally format the Terraform Plan result as part of a [GitHub Workflow Job Summary](https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/).
```

This action is intended to be used with [Atmos Terraform Apply](/integrations/github-actions/atmos-terraform-apply)

## Usage Example

```yaml
name: "atmos-terraform-plan"

on:
  workflow_dispatch:
  pull_request:
    types:
      - opened
      - synchronize
      - reopened
    branches:
      - main

# These permissions are required for GitHub to assume roles in AWS
permissions:
  id-token: write
  contents: read

jobs:
  plan:
    runs-on: ubuntu-latest
    steps:
      - name: Plan Atmos Component
        uses: cloudposse/github-action-atmos-terraform-plan@v1
        with:
          component: "foobar"
          stack: "plat-ue2-sandbox"
          component-path: "components/terraform/s3-bucket"
          terraform-plan-role: "arn:aws:iam::111111111111:role/acme-core-gbl-identity-gitops"
          terraform-state-bucket: "acme-core-ue2-auto-gitops"
          terraform-state-role: "arn:aws:iam::999999999999:role/acme-core-ue2-auto-gitops-gha"
          terraform-state-table: "acme-core-ue2-auto-gitops"
          aws-region: "us-east-2"

```

### Requirements

This GitHub Action expects an S3 bucket, DynamoDB table, and two access roles. 

#### S3 Bucket

> Any basic S3 Bucket can be connected to this action. For example, using Atmos stack configuration, define a bucket using the [`s3-bucket` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/s3-bucket) with this catalog configuration:

Reviewer: ```suggestion
Any basic S3 Bucket can be connected to this action. For example, [vendor in](/core-concepts/components/vendoring) the [`s3-component`](https://docs.cloudposse.com/components/library/aws/s3-bucket/), then using [Atmos stack configuration](/core-concepts/stacks/), define a bucket using the [`s3-bucket` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/s3-bucket) with this catalog configuration:
```

```yaml
import:
  - catalog/s3-bucket/defaults

components:
  terraform:
    # S3 Bucket for storing Terraform Plans
    gitops/s3-bucket:
      metadata:
        component: s3-bucket
        inherits:
          - s3-bucket/defaults
      vars:
        name: gitops-plan-storage
        allow_encrypted_uploads_only: false
```

Assign this S3 Bucket ARN to the `terraform-plan-bucket` input.

#### DynamoDB Table

> Similiarly, a basic DynamoDB table can be created with our [`dynamodb` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/dynamodb). Set the Hash Key and Range Key as follows:

Reviewer: ```suggestion
Similarly, a simple DynamoDB table can be provisioned using our [`dynamodb` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/dynamodb). Set the **Hash Key** and **Range Key** as follows:
```

```yaml
import:
  - catalog/dynamodb/defaults

components:
  terraform:
    # DynamoDB table used to store metadata for Terraform Plans
    gitops/dynamodb:
      metadata:
        component: dynamodb
        inherits:
          - dynamodb/defaults
      vars:
        name: gitops-plan-storage
        # These keys (case-sensitive) are required for the cloudposse/github-action-terraform-plan-storage action
        hash_key: id
        range_key: createdAt
```

> Assign this table ARN to the `terraform-plan-table` input.

Reviewer: ```suggestion
Assign this table ARN to the `terraform-plan-table` input of the `cloudposse/github-action-atmos-terraform-plan` GitHub Action.
```

#### IAM Access Roles

> First create an access role for storing and retrieving planfiles from the S3 Bucket and DynamoDB table. Assign this role ARN to the `terraform-state-role` input.

Reviewer: How to do this?

Author: added note, but link depends on https://github.com/cloudposse/terraform-aws-components/pull/735

```yaml
name: "atmos-terraform-plan"

on:
  workflow_dispatch:
  pull_request:
    types:
      - opened
      - synchronize
      - reopened
    branches:
      - main

# These permissions are required for GitHub to assume roles in AWS
permissions:
  id-token: write
  contents: read

jobs:
  plan:
    runs-on: ubuntu-latest
    steps:
      - name: Plan Atmos Component
        uses: cloudposse/github-action-atmos-terraform-plan@v1
        with:
          component: "foobar"
          stack: "plat-ue2-sandbox"
          component-path: "components/terraform/s3-bucket"
          terraform-plan-role: "arn:aws:iam::111111111111:role/acme-core-gbl-identity-gitops"
          terraform-state-bucket: "acme-core-ue2-auto-gitops"
          terraform-state-role: "arn:aws:iam::999999999999:role/acme-core-ue2-auto-gitops-gha"
          terraform-state-table: "acme-core-ue2-auto-gitops"
          aws-region: "us-east-2"

```

### Requirements

This GitHub Action expects an S3 bucket, DynamoDB table, and two access roles. 

#### S3 Bucket

This action can use any S3 Bucket to keep track of your planfiles. Just ensure the bucket is properly locked down since planfiles may contain secrets.

> For example, [vendor in](/core-concepts/components/vendoring) the [`s3-component`](https://docs.cloudposse.com/components/library/aws/s3-bucket/), then using [Atmos stack configuration](/core-concepts/stacks/), define a bucket using the [`s3-bucket` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/s3-bucket) with this catalog configuration:

Reviewer: ```suggestion
For example, [vendor in](/core-concepts/components/vendoring) the [`s3-component`](https://docs.cloudposse.com/components/library/aws/s3-bucket/), then using an [Atmos stack configuration](/core-concepts/stacks/), define a bucket using the [`s3-bucket` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/s3-bucket) with this catalog configuration:
```

```yaml
import:
  - catalog/s3-bucket/defaults

components:
  terraform:
    # S3 Bucket for storing Terraform Plans
    gitops/s3-bucket:
      metadata:
        component: s3-bucket
        inherits:
          - s3-bucket/defaults
      vars:
        name: gitops-plan-storage
        allow_encrypted_uploads_only: false
```

Assign this S3 Bucket ARN to the `terraform-plan-bucket` input.

#### DynamoDB Table

> Similarly, a simple DynamoDB table can be provisioned using our [`dynamodb` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/dynamodb). Set the **Hash Key** and **Range Key** as follows:

Reviewer: ```suggestion
Similarly, a simple DynamoDB table can be provisioned using our [`dynamodb` component](https://docs.cloudposse.com/components/library/aws/dynamodb/). Set the **Hash Key** and **Range Key** as follows:
```

---
title: Atmos Terraform Plan
sidebar_position: 40
sidebar_label: Terraform Plan
---

> The Cloud Posse GitHub Action for "Atmos Terraform Plan" runs Terraform entirely from GitHub Action workflows.

Reviewer: ```suggestion
The Cloud Posse GitHub Action for "Atmos Terraform Plan" simplifies provisioning Terraform from within GitHub using workflows. Understand precisely what to expect from running a `terraform plan` from directly within the GitHub UI for any Pull Request.
```

Given any component and stack in an Atmos supported infrastructure environment, [`github-action-atmos-terraform-plan`](https://github.com/cloudposse/github-action-atmos-terraform-plan) will run `atmos terraform plan`, generate a Terraform [planfile](https://developer.hashicorp.com/terraform/tutorials/automation/automate-terraform), store this planfile in an S3 Bucket with metadata in DynamodDB, and finally format the Terraform Plan result as part of a [GitHub Workflow Job Summary](https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/).

> This action is intended to be used with [Atmos Terraform Apply](/integrations/github-actions/atmos-terraform-apply)

Reviewer: ```suggestion
This action is intended to be used together with [Atmos Terraform Apply](/integrations/github-actions/atmos-terraform-apply) and [Atmos Affected Stacks](/integrations/github-actions/affected-stacks).
```

## Features

This GitHub Action incorporates superior GitOps support for Terraform by utilizing the capabilities of Atmos, enabling efficient management of large enterprise-scale environments.

* **Implements Native GitOps** with Atmos and Terraform
* **No hardcoded credentials.** Use GitHub OIDC to assume roles.
* **Compatible with GitHub Cloud & Self-hosted Runners** for maximum flexibility. 
* **Beautiful Job Summaries** don't clutter up pull requests with noisy GitHub comments
* **100% Open Source with Permissive APACHE2 License** means you have no expensive subscriptions or long-term commitments.

![Example Image](/img/github-actions/tf_plan.png)

![Example Image Expanded](/img/github-actions/tf_plan_expanded.png)

## Usage Example

> 

Reviewer: In this example, the action is triggered when certain events occur, such as a manual workflow dispatch or the opening, synchronization, or reopening of a pull request, specifically on the main branch. It specifies specific permissions related to assuming roles in AWS. Within the "plan" job, the "component" and "stack" are hardcoded. In practice, these are usually derived from another action.

> 

Reviewer: ```suggestion
In this example, the action is triggered when certain events occur, such as a manual workflow dispatch or the opening, synchronization, or reopening of a pull request, specifically on the main branch. It specifies specific permissions related to assuming roles in AWS. Within the "plan" job, the "component" and "stack" are hardcoded. In practice, these are usually derived from another action. 
```

```yaml
name: "atmos-terraform-plan"

on:
  workflow_dispatch:
  pull_request:
    types:
      - opened
      - synchronize
      - reopened
    branches:
      - main

# These permissions are required for GitHub to assume roles in AWS
permissions:
  id-token: write
  contents: read

jobs:
  plan:
    runs-on: ubuntu-latest
    steps:
      - name: Plan Atmos Component
        uses: cloudposse/github-action-atmos-terraform-plan@v1
        with:
          component: "foobar"
          stack: "plat-ue2-sandbox"
          component-path: "components/terraform/s3-bucket"
          terraform-plan-role: "arn:aws:iam::111111111111:role/acme-core-gbl-identity-gitops"
          terraform-state-bucket: "acme-core-ue2-auto-gitops"
          terraform-state-role: "arn:aws:iam::999999999999:role/acme-core-ue2-auto-gitops-gha"
          terraform-state-table: "acme-core-ue2-auto-gitops"
          aws-region: "us-east-2"

> ```
> 

Reviewer: ```suggestion
```

Please note that in practice, we recommend combining this action with the [`affected-stacks`](/integrations/github-actions/affected-stacks) GitHub Action inside a matrix to plan all affected stacks in parallel.
```

### Requirements

> This GitHub Action expects an S3 bucket, DynamoDB table, and two access roles. 

Reviewer: ```suggestion
This GitHub Action depends on a few resources:
* **S3 bucket** for storing planfiles
* **DynamoDB table** for retrieving metadata about planfiles
* **2x IAM roles** for "planning" and accessing the "state" bucket
```

#### S3 Bucket

This action can use any S3 Bucket to keep track of your planfiles. Just ensure the bucket is properly locked down since planfiles may contain secrets.

For example, [vendor in](/core-concepts/components/vendoring) the [`s3-component`](https://docs.cloudposse.com/components/library/aws/s3-bucket/), then using an [Atmos stack configuration](/core-concepts/stacks/), define a bucket using the [`s3-bucket` component](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/s3-bucket) with this catalog configuration:

```yaml
import:
  - catalog/s3-bucket/defaults

components:
  terraform:
    # S3 Bucket for storing Terraform Plans
    gitops/s3-bucket:
      metadata:
        component: s3-bucket
        inherits:
          - s3-bucket/defaults
      vars:
        name: gitops-plan-storage

>         allow_encrypted_uploads_only: false

Author: this planfiles are not encrypted when push/pulled with https://github.com/cloudposse/github-action-terraform-plan-storage

>         allow_encrypted_uploads_only: false

Reviewer: From what I can tell, it's using SSE (Server Side Encryption)

https://github.com/cloudposse/github-action-terraform-plan-
storage/blob/68f48c4cab5d6163d634df904ef654b7e69c4e1f/src/lib/s3/s3.ts#L17-L22

>         allow_encrypted_uploads_only: false

Reviewer: We will resolve this in a follow up PR.

```

Assign this S3 Bucket ARN to the `terraform-plan-bucket` input.

#### DynamoDB Table

Similarly, a simple DynamoDB table can be provisioned using our [`dynamodb` component](https://docs.cloudposse.com/components/library/aws/dynamodb/). Set the **Hash Key** and **Range Key** as follows:

```yaml
import:
  - catalog/dynamodb/defaults

components:
  terraform:
    # DynamoDB table used to store metadata for Terraform Plans
    gitops/dynamodb:
      metadata:
        component: dynamodb
        inherits:
          - dynamodb/defaults
      vars:
        name: gitops-plan-storage
        # These keys (case-sensitive) are required for the cloudposse/github-action-terraform-plan-storage action
        hash_key: id
        range_key: createdAt
```

> 

Reviewer: ```suggestion
```

> Pass the ARN of this table as the input to the `terraform-plan-table` of the `cloudposse/github-action-atmos-terraform-plan` GitHub Action.

Reviewer: ```suggestion
Pass the ARN of this table as the input to the `terraform-plan-table` of the [`cloudposse/github-action-atmos-terraform-plan`](https://github.com/cloudposse/github-action-atmos-terraform-plan) GitHub Action.
```
